boa_engine/src/error.rs

```diff
@@ -0,0 +1,787 @@
+//! Error-related types and conversions.
+
+use crate::{
+    builtins::{error::ErrorKind, Array},
+    object::JsObject,
+    object::ObjectData,
+    property::PropertyDescriptor,
+    syntax::parser,
+    Context, JsValue,
+};
+use boa_gc::{Finalize, Trace};
+use flexstr::LocalStr;
+use std::{borrow::Cow, fmt::Display};
+use thiserror::Error;
+
+/// The error type returned by all operations related
+/// to the execution of Javascript code.
+///
+/// This is essentially an enum that can store either [`JsNativeError`]s (for ideal
+/// native errors)  or opaque [`JsValue`]s, since Javascript allows throwing any valid
+/// `JsValue`.
+///
+/// The implementation doesn't provide a [`From`] conversion
+/// for `JsValue`. This is with the intent of encouraging the usage of proper
+/// `JsNativeError`s instead of plain `JsValue`s. However, if you
+/// do need a proper opaque error, you can construct one using the
+/// [`JsError::from_opaque`] method.
+///
+/// # Examples
+///
+/// ```rust
+/// # use boa_engine::{JsError, JsNativeError, JsNativeErrorKind, JsValue};
+/// let cause = JsError::from_opaque("error!".into());
+///
+/// assert!(cause.as_opaque().is_some());
+/// assert_eq!(cause.as_opaque().unwrap(), &JsValue::from("error!"));
+///
+/// let native_error: JsError = JsNativeError::typ()
+///     .with_message("invalid type!")
+///     .with_cause(cause)
+///     .into();
+///
+/// assert!(native_error.as_native().is_some());
+///
+/// let kind = &native_error.as_native().unwrap().kind;
+/// assert!(matches!(kind, JsNativeErrorKind::Type));
+/// ```
+#[derive(Debug, Clone, Trace, Finalize)]
+pub struct JsError {
+    inner: Repr,
+}
+
+// hiding the enum makes it a bit more difficult
+// to treat `JsError` as a proper enum, which should
+// make it a bit harder to try to match against a native error
+// without calling `try_native` first.
+#[derive(Debug, Clone, Trace, Finalize)]
+enum Repr {
+    Native(JsNativeError),
+    Opaque(JsValue),
+}
```

Comment:
Even if we want to hide it, I would add some documentation explaining its usage, with triple slashes: `///`

We could also add this to that documentation.